doc/specs/thrift-uri.md

```diff
@@ -0,0 +1,155 @@
+Thrift URIs
+====================================================================
+
+Last Modified: 2024-APR-21
+
+<!--
+--------------------------------------------------------------------
+
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements. See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership. The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License. You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied. See the License for the
+specific language governing permissions and limitations
+under the License.
+
+--------------------------------------------------------------------
+-->
+
+# Motivation and use case
+
+Describing the endpoint specifics for a Thrift API today is a purely text-adventure like exercise, which leaves the client side developer end with the tedious task to set up and put together a proper protocol/transport stack. That sometimes even leads to headaches, e.g. if the server requires e.g. framed protocol which might not be obvious. The use of a generally accepted, machine-readable and extensible Thrift URI format to describe client bindings could streamline that process.
```

Comment:
if we limit the scope of the issue to endpoints (e.g. limit to RPC part, and exclude pure serialization/deserialization use-cases), I would say the THeaderProtocol resolves the _majority_ of the problems here.

THeaderProtocol already supports the auto detection of:

* binary vs. compact
* framed or not

it also supports zlib (but only if it's enabled inside Header, it does not support raw zlib transport unlike raw framed/binary/compact)

So the only issues left are:

* json protocol is not covered (supported)
* the underlying, final, raw transport used (socket/sslsocket/file/pipes/namedpipes)

(I think memory and buffered used in RPCs are pure implementation details and do not really affect the wire protocol, but please correct me if I'm wrong)

so maybe we should shift focus to support THeader in more language implementations (it's currently only implemented in c++/py/go)?

when a server uses THeaderProtocol, any of the following client will be able to talk to it without issue:

* THeaderProtocol
* TBinaryProtocol
* TBinaryProtocol+TFramedTransport
* TCompactProtocol
* TCompactProtocol+TFramedTransport

(the final raw transport, e.g. one of socket/sslsocket/file/pipe/namedpipe, still need to be specified)

that won't resolve the json problem, obvious (I also don't think add json support to THeader is a good idea), but from my experience TJSONProtocol used in thrift RPC are usually more special cases (for example, only being combined with THTTPTransport) that it's _usually_ unambiguous and has much less chance for the client to use the wrong protocol/transport.

Reply:
> So the only issues left are:

* Extensibility: How do I integrate my custom transport or protocol into the THeader system?
* How does THeader deal with multiplex?

> memory and buffered used in RPCs are pure implementation details and do not really affect the wire protocol

Kind of "for the sake of completeness" thing. They probably could be safely excluded.

> As referenced in https://www.ietf.org/rfc/rfc2396.txt, even though this is not how URI is defined, a common syntax used widely defines it as:

Yes I noticed. [RFC3986](https://www.ietf.org/rfc/rfc3986.txt) is more recent, but essentially tells the same story. Will be fixed soon.